x-pack/plugins/security_solution/server/lib/detection_engine/signals/types.ts

```diff
@@ -179,3 +189,39 @@ export interface QueryFilter {
     must_not: Filter[];
   };
 }
+
+export interface SearchAfterAndBulkCreateParams {
+  gap: moment.Duration | null;
+  previousStartedAt: Date | null | undefined;
+  ruleParams: RuleTypeParams;
+  services: AlertServices;
+  listClient: ListClient;
+  exceptionsList: ExceptionListItemSchema[];
+  logger: Logger;
+  id: string;
+  inputIndexPattern: string[];
+  signalsIndex: string;
+  name: string;
+  actions: RuleAlertAction[];
+  createdAt: string;
+  createdBy: string;
+  updatedBy: string;
+  updatedAt: string;
+  interval: string;
+  enabled: boolean;
+  pageSize: number;
+  filter: unknown;
+  refresh: RefreshTypes;
+  tags: string[];
+  throttle: string;
+  buildRuleMessage: BuildRuleMessage;
+}
+
+export interface SearchAfterAndBulkCreateReturnType {
+  success: boolean;
+  searchAfterTimes: string[];
+  bulkCreateTimes: string[];
+  lastLookBackDate: Date | null | undefined;
```

Comment:
It is a bit weird. It sometimes can be undefined if it has never been set before. I think we used `null` as setting it for another type of value. But yeah, it can be either of those three states.